gr-video-sdl/lib/sink_uc_impl.cc

```diff
@@ -226,67 +119,142 @@ int sink_uc_impl::work(int noutput_items,
+                NULL);
             src_pixels_0 += d_chunk_size;
         }
         break;
     default: // 0 or more then 3 channels
-        d_logger->error(
-            "Wrong number of channels: 1, 2 or 3 channels are supported. Requested "
-            "number of channels is {:d}",
-            input_items.size());
-        throw std::runtime_error("video_sdl::sink_uc");
+        d_logger->error("Wrong number of channels: 1, 2 or 3 channels are supported. "
+                        "Requested number of channels is {:d}",
+                        input_items.size());
+        throw std::runtime_error("video_sdl2::sink_uc");
     }
 
-    SDL_UnlockYUVOverlay(d_image);
     return noutput_items_produced;
 }
 
+int render_loop_uc(void* data)
+{
+    auto sink = reinterpret_cast<sink_uc_impl*>(data);
+
+    if (SDL_Init(SDL_INIT_VIDEO) < 0) {
+        sink->d_logger->error(
+            "Couldn't initialize SDL: {:s}; SDL_Init(SDL_INIT_VIDEO) failed",
+            SDL_GetError());
+        throw std::runtime_error("video_sdl2::sink_uc::render_loop");
+    }
+
+    /* accept any depth */
+    auto display = SDL_SetVideoMode(sink->d_dst_width,
+                                    sink->d_dst_height,
+                                    0,
+                                    SDL_SWSURFACE | SDL_RESIZABLE | SDL_ANYFORMAT);
+    // SDL_DOUBLEBUF |SDL_SWSURFACE| SDL_HWSURFACE||SDL_FULLSCREEN
+    if (!display) {
+        sink->d_logger->error("Unable to set SDL video mode: {:s}", SDL_GetError());
+        throw std::runtime_error("video_sdl2::sink_uc::render_loop");
+    }
+
+    auto overlay =
+        SDL_CreateYUVOverlay(sink->d_width, sink->d_height, SDL_IYUV_OVERLAY, display);
+    if (!overlay) {
+        sink->d_logger->error("Couldn't create a YUV overlay: {:s}", SDL_GetError());
+        throw std::runtime_error("video_sdl2::sink_uc::render_loop");
+    }
+
+    SDL_Rect dstrect{ 0, 0, (Uint16)sink->d_dst_width, (Uint16)sink->d_dst_height };
+
+    while (!SDL_QuitRequested()) {
+        if (sink->d_wanted_ticks == 0)
+            sink->d_wanted_ticks = SDL_GetTicks();
+
+        if (sink->d_avg_delay > 0)
+            SDL_Delay((unsigned int)sink->d_avg_delay);
+
+        if (SDL_LockYUVOverlay(overlay) < 0) {
+            continue;
+        }
+        for (int i = 0; i < sink->d_height; i++) {
+            memcpy(&overlay->pixels[0][i * overlay->pitches[0]],
+                   &sink->d_buf_y[i * sink->d_width],
+                   sink->d_width);
+            memcpy(&overlay->pixels[1][i * overlay->pitches[1]],
+                   &sink->d_buf_u[i * sink->d_width / 2],
+                   sink->d_width / 2);
+            memcpy(&overlay->pixels[2][i * overlay->pitches[2]],
+                   &sink->d_buf_v[i * sink->d_width / 2],
+                   sink->d_width / 2);
+        }
+        SDL_UnlockYUVOverlay(overlay);
+        SDL_DisplayYUVOverlay(overlay, &dstrect);
+    }
+
+    SDL_FreeYUVOverlay(overlay);
+    exit(0);
```

Comment:
Can't exit() in a block.

Reply:
Is there a way to broadcast a "stop" signal not just to the block, but to the whole flowgraph?

Reply:
A display wouldn't typically want to terminate the flowgraph. If a work() function returns DONE, then the framework does not expect it to take any more data, and that would result in the flowgraph stopping (eventually?).